File: haskell/src/KMeansStuff.hs
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE FlexibleContexts #-}

module KMeansStuff where

import Data.Metric (Metric(..))
import Data.Vector (Vector(..))
import qualified Data.Vector as V
import Algorithms.Lloyd.Sequential (Point(..), Cluster(..), ExpectDivergent(..))
import Algorithms.Lloyd.Strategies (Partitions(..))
import qualified Algorithms.Lloyd.Sequential as LloydSeq
import qualified Algorithms.Lloyd.Strategies as LloydPar
import Data.Function (on)
import Testing (Parallelism(..), Seed, Size, Interval, Dimensions, vectorFromSeed)
import Util ((<$$>))
import Data.Random (Uniform, Distribution, uniform, sample)

type K = Int

instance Metric (Vector Double) where
  distance = squareDistance

squareDistance :: Num a => Vector a -> Vector a -> a
squareDistance = V.sum <$$> V.zipWith ((^2) <$$> (-))

pointFromSeed :: (Real a, Distribution Uniform a) => Seed -> Interval a -> Dimensions -> Point
pointFromSeed seed lohi dim = Point $ fmap realToFrac $ vectorFromSeed seed lohi dim

generateKMeansData :: (Real a, Distribution Uniform a) => Seed -> (a, a) -> Dimensions -> Size -> Vector Point
generateKMeansData seed lohi dim n = V.map (Point . fmap realToFrac . (flip (flip vectorFromSeed lohi) dim) . fromIntegral) (V.fromList [1..n])

runKmeans :: Parallelism Int -> Vector Point -> K -> Vector (Vector Point)
runKmeans parallelism points k = kmeans expectDivergent metric points initial
  where
    kmeans = case parallelism of
      Sequential -> LloydSeq.kmeans
      Parallel p -> ($ Partitions p) <$$> LloydPar.kmeans
    initial :: Vector Cluster
    initial =
      if V.length points < k
      then error $ show (V.length points) ++ " points, but k = " ++ show k
      else V.zipWith (flip Cluster) points (V.fromList [0 .. k-1])
    expectDivergent :: ExpectDivergent
    expectDivergent = ExpectDivergent 100000
    metric :: Vector Double -> Vector Double
    metric = id

File: scala/src/main/scala/programs/KMeansStuff.scala
import Lloyd.Sequential._
import Lloyd.Sequential.{kmeans => kmeansSeq}
import Lloyd.Strategies.{kmeans => kmeansPar}

package object KMeansStuff {

  def squareDistance(a: Point, b: Point): Double = {
    val square = (x: Double) => x*x
    val subtract = (a: Double, b: Double) => a - b
    a.zip(b).map(subtract.tupled andThen square).sum
  }
  
  val expectDivergent = 100000

  def runKMeans(points: Vector[Point], k: Int, parallelism: Int): Vector[Vector[Point]] = {
    val initial =
      if (points.length < k)
        throw new Error(points.length + " points, but k = " + k)
      else
        points.take(k).zipWithIndex.map(tup => Cluster(tup._2, tup._1))
    if (parallelism > 1)
      kmeansPar(expectDivergent, squareDistance, parallelism, points, initial)
    else
      kmeansSeq(expectDivergent, squareDistance, points, initial)
  }

}
